File: tests.php5/callback.php
#!/usr/bin/php

<?php
include_once ("java/Java.inc");
function c1($s1, $o1) {
  echo "c1: $s1, $o1\n";
  // must display a php warning
  throw new JavaException("java.lang.Exception", "bleh!");

  // not reached 
  echo "ERROR.\n"; 
  exit(3); 
}
function c2($b) {
  echo "c2: $b\n";
  return $b;
}
function c3 ($e) {
  echo "c3: $e\n";
  return 2;
}
$here=realpath(dirname($_SERVER["SCRIPT_FILENAME"]));
if(!$here) $here=getcwd();
java_require("$here/callback.jar");

$closure=java_closure(null, null, java("Callback"));
$callbackTest=new java('Callback$Test', $closure);

if($callbackTest->test()) {
  echo "test okay\n";
  exit(0);
} else {
  echo "test failed\n";
  exit(1);
}
?>

File: tests.php5/Callback.java
public interface Callback {
    public void c1(String s1, Object o1) throws Exception;
    public boolean c2(boolean b);
    public long c3(Exception e);
    public long ni(Exception e);

    public class Test {
	public Callback cb;
	public Test(Callback closure) {
	    cb=closure;
	}
	public long test1() {
	    try {
		cb.c1(new String("TEST"), this);
		return 1;
	    } catch (Exception e) {
		return cb.c3(e);
	    }
	}
	public boolean test() {
	    return cb.c2(test1()==2);
	}
    }
}
    
